### PR TITLE
Add Tags, Default Install Url, Scopes and Permissions to ApplicationInfo

### PR DESCRIPTION
**Pull request by Xirado**
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

### Changes

- [x] Internal code
- [x] Library interface (affecting end-user code) 
- [x] Documentation
- [ ] Other: \_____ 

Closes Issue: NaN

## Description

This pull request adds `getTags()`, `getCustomAuthorizationUrl()`, `getPermissions()` and `getScopes()` to `ApplicationInfo`

### CODE REVIEW

File: src/main/java/net/dv8tion/jda/internal/entities/EntityBuilder.java [hunk 1]
```diff
@@ -2099,9 +2100,22 @@ public ApplicationInfo createApplicationInfo(DataObject object)
...
         final User owner = createUser(object.getObject("owner"));
         final ApplicationTeam team = !object.isNull("team") ? createApplicationTeam(object.getObject("team")) : null;
+        final String customAuthUrl = object.getString("custom_install_url", null);
+        final List<String> tags = !object.isNull("tags")
+                ? Collections.unmodifiableList(object.getArray("tags").stream(DataArray::getString).collect(Collectors.toList()))
+                : Collections.emptyList();
+
+        final long defaultAuthUrlPerms = !object.isNull("install_params")
+                ? Long.parseLong(object.getObject("install_params").getString("permissions"))
+                : 0L;
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/internal/entities/EntityBuilder.java:2110]:
```suggestion
        final long defaultAuthUrlPerms = object.optObject("install_params").map(o -> o.getLong("permissions")).orElse(0);
```

File: src/main/java/net/dv8tion/jda/internal/entities/EntityBuilder.java [hunk 2]
```diff
@@ -2099,9 +2100,22 @@ public ApplicationInfo createApplicationInfo(DataObject object)
         final boolean isBotPublic = object.getBoolean("bot_public");
         final User owner = createUser(object.getObject("owner"));
         final ApplicationTeam team = !object.isNull("team") ? createApplicationTeam(object.getObject("team")) : null;
+        final String customAuthUrl = object.getString("custom_install_url", null);
+        final List<String> tags = !object.isNull("tags")
+                ? Collections.unmodifiableList(object.getArray("tags").stream(DataArray::getString).collect(Collectors.toList()))
+                : Collections.emptyList();
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/internal/entities/EntityBuilder.java:2106]:
```suggestion
        final List<String> tags = object.optArray("tags").orElseGet(DataArray::empty())
                    .stream(DataArray::getString)
                    .collect(Collectors.toList()));
```

File: src/main/java/net/dv8tion/jda/internal/entities/ApplicationInfoImpl.java [hunk 3]
```diff
@@ -56,6 +63,11 @@ public ApplicationInfoImpl(JDA api, String description, boolean doesBotRequireCo
         this.privacyPolicyUrl = privacyPolicyUrl;
         this.owner = owner;
         this.team = team;
+        this.tags = tags;
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/internal/entities/ApplicationInfoImpl.java:66]:
```suggestion
        this.tags = Collections.unmodifiableList(tags);
```

File: src/main/java/net/dv8tion/jda/internal/entities/ApplicationInfoImpl.java [hunk 4]
```diff
@@ -56,6 +63,11 @@ public ApplicationInfoImpl(JDA api, String description, boolean doesBotRequireCo
...
+        this.customAuthUrl = customAuthUrl;
+        this.defaultAuthUrlPerms = defaultAuthUrlPerms;
+        this.defaultAuthUrlScopes = defaultAuthUrlScopes;
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/internal/entities/ApplicationInfoImpl.java:69]:
```suggestion
        this.defaultAuthUrlScopes = Collection.unmodifiableList(defaultAuthUrlScopes);
```

File: src/main/java/net/dv8tion/jda/internal/entities/EntityBuilder.java [hunk 5]
```diff
@@ -2099,9 +2100,22 @@ public ApplicationInfo createApplicationInfo(DataObject object)
...
+
+        final long defaultAuthUrlPerms = !object.isNull("install_params")
+                ? Long.parseLong(object.getObject("install_params").getString("permissions"))
+                : 0L;
+
+        final List<String> defaultAuthUrlScopes = !object.isNull("install_params")
+                ? Collections.unmodifiableList(object.getObject("install_params").getArray("scopes").stream(DataArray::getString).collect(Collectors.toList()))
+                : Collections.emptyList();
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/internal/entities/EntityBuilder.java:2114]:
```suggestion
        final List<String> defaultAuthUrlScopes = object.optObject("install_params").map(obj ->
                    obj.getArray("scopes").stream(DataArray::getString).collect(Collectors.toList()))
        ).orElse(Collections.emptyList());
```

File: src/main/java/net/dv8tion/jda/api/entities/ApplicationInfo.java [hunk 6]
```diff
@@ -287,4 +286,44 @@ default String getInviteUrl(long guildId, @Nullable Permission... permissions)
...
+
+    /**
+     * A {@link java.util.EnumSet} of permissions the default authorization URL is set up with.
+     *
+     * This is empty if you set a custom URL in the Developer Portal.
+     *
+     * @return Set of permissions the default authorization URL is set up with.
+     */
+    @Nonnull
+    EnumSet<Permission> getPermissions();
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/entities/ApplicationInfo.java:328]: Add `getPermissionsRaw(): Long`

File: src/main/java/net/dv8tion/jda/api/entities/ApplicationInfo.java [hunk 7]
```diff
@@ -287,4 +286,44 @@ default String getInviteUrl(long guildId, @Nullable Permission... permissions)
...
+     *
+     * @return List of scopes the default authorization URL is set up with.
+     */
+    @Nonnull
+    List<String> getScopes();
+
+    /**
+     * A {@link java.util.EnumSet} of permissions the default authorization URL is set up with.
+     *
+     * This is empty if you set a custom URL in the Developer Portal.
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/entities/ApplicationInfo.java:323]:
```suggestion
     * <p>This is empty if you set a custom URL in the <a href="https://discord.com/developers/applications" target="_blank">Developer Portal</a>.
```

File: src/main/java/net/dv8tion/jda/api/entities/ApplicationInfo.java [hunk 8]
```diff
@@ -287,4 +286,44 @@ default String getInviteUrl(long guildId, @Nullable Permission... permissions)
...
+     *
+     * @return Custom Authorization URL, or null if it has not been set
+     */
+    @Nullable
+    String getCustomAuthorizationUrl();
+
+    /**
+     * A {@link java.util.List} of scopes the default authorization URL is set up with.
+     *
+     * This is empty if you set a custom URL in the Developer Portal.
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/entities/ApplicationInfo.java:313]:
```suggestion
     * <p>This is empty if you set a custom URL in the <a href="https://discord.com/developers/applications" target="_blank">Developer Portal</a>.
```

File: src/main/java/net/dv8tion/jda/api/entities/ApplicationInfo.java [hunk 9]
```diff
@@ -287,4 +286,44 @@ default String getInviteUrl(long guildId, @Nullable Permission... permissions)
...
+     *
+     * @return List containing the tags of this bot's application
+     */
+    @Nonnull
+    List<String> getTags();
+
+    /**
+     * The custom Authorization URL of this bot's application.
+     *
+     * This returns null if no custom URL is set in the Developer Portal or if In-app Authorization is enabled.
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/entities/ApplicationInfo.java:303]:
```suggestion
     * <p>This returns null if no custom URL is set in the <a href="https://discord.com/developers/applications" target="_blank">Developer Portal</a> or if In-app Authorization is enabled.
```

File: src/main/java/net/dv8tion/jda/api/entities/ApplicationInfo.java [hunk 10]
```diff
@@ -287,4 +286,44 @@ default String getInviteUrl(long guildId, @Nullable Permission... permissions)
      * @return Whether the bot is public
      */
     boolean isBotPublic();
+
+    /**
+     * A {@link java.util.List} containing the tags of this bot's application.
+     *
+     * This List is empty if no tags are set in the Developer Portal.
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/entities/ApplicationInfo.java:293]:
```suggestion
     * <p>This List is empty if no tags are set in the <a href="https://discord.com/developers/applications" target="_blank">Developer Portal</a>.
```

File: src/main/java/net/dv8tion/jda/api/entities/ApplicationInfo.java [hunk 11]
```diff
@@ -287,4 +286,44 @@ default String getInviteUrl(long guildId, @Nullable Permission... permissions)
...
+     *
+     * This is empty if you set a custom URL in the Developer Portal.
+     *
+     * @return List of scopes the default authorization URL is set up with.
+     */
+    @Nonnull
+    List<String> getScopes();
+
+    /**
+     * A {@link java.util.EnumSet} of permissions the default authorization URL is set up with.
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/entities/ApplicationInfo.java:321]:
```suggestion
     * An {@link java.util.EnumSet} of permissions the default authorization URL is set up with.
```

File: src/main/java/net/dv8tion/jda/internal/entities/ApplicationInfoImpl.java [hunk 12]
```diff
@@ -40,10 +43,14 @@
     private final String name;
     private final User owner;
     private final ApplicationTeam team;
+    private final List<String> tags;
+    private final String customAuthUrl;
+    private final EnumSet<Permission> defaultAuthUrlPerms;
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/internal/entities/ApplicationInfoImpl.java:48]: These should be stored as `long`, there isn't really a reason to parse them unless the user needs it.

File: src/main/java/net/dv8tion/jda/api/entities/ApplicationInfo.java [hunk 13]
```diff
@@ -287,4 +286,44 @@ default String getInviteUrl(long guildId, @Nullable Permission... permissions)
...
+     */
+    @Nullable
+    String getCustomAuthorizationUrl();
+
+    /**
+     * A {@link java.util.List} of scopes the default authorization URL is set up with.
+     *
+     * This is empty if you set a custom URL in the Developer Portal.
+     *
+     * @return List of scopes the default authorization URL is set up with.
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/entities/ApplicationInfo.java:315]:
```suggestion
     * @return Immutable list of scopes the default authorization URL is set up with.
```

File: src/main/java/net/dv8tion/jda/api/entities/ApplicationInfo.java [hunk 14]
```diff
@@ -287,4 +286,44 @@ default String getInviteUrl(long guildId, @Nullable Permission... permissions)
      * @return Whether the bot is public
      */
     boolean isBotPublic();
+
+    /**
+     * A {@link java.util.List} containing the tags of this bot's application.
+     *
+     * This List is empty if no tags are set in the Developer Portal.
+     *
+     * @return List containing the tags of this bot's application
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/entities/ApplicationInfo.java:295]:
```suggestion
     * @return Immutable list containing the tags of this bot's application
```

**Comment by Xirado** [comment]
[DDocs PR](https://github.com/discord/discord-api-docs/pull/4112) has been merged

File: src/main/java/net/dv8tion/jda/api/entities/ApplicationInfo.java [hunk 15]
```diff
@@ -287,4 +286,51 @@ default String getInviteUrl(long guildId, @Nullable Permission... permissions)
...
+     *
+     * @return Immutable list containing the tags of this bot's application
+     */
+    @Nonnull
+    List<String> getTags();
+
+    /**
+     * The custom Authorization URL of this bot's application.
+     *
+     * <p>This returns null if no custom URL is set in the <a href="https://discord.com/developers/applications" target="_blank">Developer Portal</a> or if In-app Authorization is enabled.
```
Comment (Chew) [inline src/main/java/net/dv8tion/jda/api/entities/ApplicationInfo.java:303]:
```suggestion
     * <p>This returns {@code null} if no custom URL is set in the <a href="https://discord.com/developers/applications" target="_blank">Developer Portal</a> or if In-app Authorization is enabled.
```

File: src/main/java/net/dv8tion/jda/api/entities/ApplicationInfo.java [hunk 16]
```diff
@@ -287,4 +286,51 @@ default String getInviteUrl(long guildId, @Nullable Permission... permissions)
...
+     *
+     * @return Custom Authorization URL, or null if it has not been set
+     */
+    @Nullable
+    String getCustomAuthorizationUrl();
+
+    /**
+     * A {@link java.util.List} of scopes the default authorization URL is set up with.
+     *
+     * <p>This is empty if you set a custom URL in the <a href="https://discord.com/developers/applications" target="_blank">Developer Portal</a>.
```
Comment (Chew) [inline src/main/java/net/dv8tion/jda/api/entities/ApplicationInfo.java:313]: Above we specify "This _List_ is empty..." do we wanna specify that here, too?

File: src/main/java/net/dv8tion/jda/internal/entities/ApplicationInfoImpl.java [hunk 17]
```diff
@@ -56,6 +64,11 @@ public ApplicationInfoImpl(JDA api, String description, boolean doesBotRequireCo
         this.privacyPolicyUrl = privacyPolicyUrl;
         this.owner = owner;
         this.team = team;
+        this.tags = Collections.unmodifiableList(tags);
+        this.customAuthUrl = customAuthUrl;
+        this.defaultAuthUrlPerms = defaultAuthUrlPerms;
+        this.defaultAuthUrlScopes = Collections.unmodifiableList(defaultAuthUrlScopes);
+
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/internal/entities/ApplicationInfoImpl.java:71]:
```suggestion
```

File: src/main/java/net/dv8tion/jda/internal/entities/EntityBuilder.java [hunk 18]
```diff
@@ -2099,9 +2099,23 @@ public ApplicationInfo createApplicationInfo(DataObject object)
...
+        final String customAuthUrl = object.getString("custom_install_url", null);
+        final List<String> tags = object.optArray("tags").orElseGet(DataArray::empty)
+                    .stream(DataArray::getString)
+                    .collect(Collectors.toList());
+
+        final long defaultAuthUrlPerms = object.optObject("install_params")
+                    .map(o -> o.getLong("permissions"))
+                    .orElse(0L);
+
+        final List<String> defaultAuthUrlScopes = object.optObject("install_params")
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/internal/entities/EntityBuilder.java:2111]: Put `object.optObject("install_params")` into a variable so it doesn't have to allocate twice

**Review by MinnDevelopment — APPROVED** [review]

**Review by DV8FromTheWorld — APPROVED** [review]

**Pull request merged**